src/operator/semijoin.rs

```diff
@@ -0,0 +1,609 @@
+use crate::{
+    algebra::{HasZero, IndexedZSet, MulByRef, ZSet},
+    circuit::{
+        operator_traits::{BinaryOperator, Data, Operator},
+        Scope,
+    },
+    lattice::Lattice,
+    trace::{ord::OrdValSpine, Batch, BatchReader, Batcher, Builder, Cursor, Trace, TraceReader},
+    Circuit, Stream, Timestamp,
+};
+use deepsize::DeepSizeOf;
+use hashbrown::HashMap;
+use std::{
+    borrow::Cow,
+    cmp::{min, Ordering},
+    fmt::Write,
+    hash::Hash,
+    marker::PhantomData,
+};
+use timely::PartialOrder;
+
+impl<S, Pairs> Stream<Circuit<S>, Pairs>
+where
+    S: Clone + 'static,
+{
+    /// Semijoin two streams of batches.
+    ///
+    /// The operator takes two streams of batches indexed with the same key type
+    /// (`Pairs::Key = Keys::Key`) and outputs a stream obtained by joining each
+    /// pair of inputs.
+    ///
+    /// Input streams will typically be produced by [`Stream::index()`] or
+    /// [`Stream::index_with()`]
+    ///
+    /// #### Type arguments
+    ///
+    /// * `Pairs` - batch type in the first input stream.
+    /// * `Keys` - batch type in the second input stream.
+    /// * `Out` - output Z-set type.
+    pub fn semijoin_stream<Keys, Out>(
+        &self,
+        keys: &Stream<Circuit<S>, Keys>,
+    ) -> Stream<Circuit<S>, Out>
+    where
+        // TODO: Associated type bounds (rust/#52662) really simplify things
+        // TODO: Allow non-unit timestamps
+        Pairs: Batch<Time = ()> + Data + Send,
+        Pairs::Key: Hash + Ord + Clone,
+        Pairs::Val: Ord + Clone,
+        Pairs::R: HasZero + MulByRef,
+        Keys: Batch<Key = Pairs::Key, Val = (), Time = (), R = Pairs::R> + Data + Send,
```

Comment:
I don't think it's strictly necessary that `Val = ()`. The second collection can have any value type.  I guess this makes it easier to deal with weights?